sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala

```diff
@@ -211,6 +211,49 @@ object GpuOrcScan extends Arm {
         } else {
           downCastAnyInteger(col, toDt)
         }
+      // bool to float, double(float64)
+      case (DType.BOOL8, DType.FLOAT32 | DType.FLOAT64) =>
+        col.castTo(toDt)
+      // bool to string
+      case (DType.BOOL8, DType.STRING) =>
+        withResource(col.castTo(toDt)) { casted =>
+          // cuDF produces "ture"/"false" while CPU outputs "TRUE"/"FALSE".
+          casted.upper()
+        }
+      // integer to float, double(float64), string
+      case (DType.INT8 | DType.INT16 | DType.INT32 | DType.INT64,
+      DType.FLOAT32 | DType.FLOAT64 | DType.STRING) =>
+        col.castTo(toDt)
+      // {bool, integer} to timestamp(micro seconds)
+      case (DType.BOOL8 | DType.INT8 | DType.INT16 | DType.INT32, DType.TIMESTAMP_MICROSECONDS) =>
+        // cuDF requires casting to a long first.
+        withResource(col.castTo(DType.INT64)) { longs =>
+          // In CPU, ORC assumes the integer value is in seconds, and returns timestamp in
+          // micro seconds.
+          withResource(longs.bitCastTo(DType.TIMESTAMP_SECONDS)) { seconds =>
+            seconds.castTo(DType.TIMESTAMP_MICROSECONDS)
+          }
+        }
+      case (DType.INT64, DType.TIMESTAMP_MICROSECONDS) =>
+        // We need overflow checking here, since max value of INT64 is about 9 * 1e18, and convert
+        // INT64 to micro seconds(also a INT64 actually), we need multiply 1e6, it may cause long
+        // overflow.
+        // If val * 1e6 / 1e6 == val, then there is no overflow.
+        withResource(ColumnVector.fromScalar(Scalar.fromLong(1000000),
```

Comment:
1 Make this overflow check be an untils method and call it here.
2 This too-many nested `withResource` calls will lead to high GPU memory peak, try to close the columns are no longer needed as soon as possible. e.g. `vec` can be closed when computing `isEqual`.

Reply:
I found that there is `max` method in `ColumnVector`. I think checking it in this way is better.
```scala
val k = 1e6.toLong
val maxVal = col.max().getLong()
if (maxVal * k / k != maxVal) {
  throw new ArithmeticException("Long overflow")
}
```